libvast/src/directory.cpp

```diff
@@ -128,32 +128,32 @@ caf::expected<size_t> recursive_size(const std::filesystem::path& root_dir) {
   return total_size;
 }
 
-std::vector<path>
-filter_dir(const path& dir, std::function<bool(const path&)> filter,
+caf::expected<std::vector<std::filesystem::path>>
+filter_dir(const std::filesystem::path& root_dir,
+           std::function<bool(const std::filesystem::path&)> filter,
            size_t max_recursion) {
-  std::vector<path> result;
-  if (max_recursion == 0)
-    return result;
-  for (auto& f : directory(dir)) {
-    switch (f.kind()) {
-      default: {
-        if (!filter || filter(f))
-          result.push_back(f);
-        break;
-      }
-      case path::directory: {
-        // Recurse directories depth-first.
-        auto paths = filter_dir(f, filter, --max_recursion);
-        if (!paths.empty()) {
-          auto begin = std::make_move_iterator(paths.begin());
-          auto end = std::make_move_iterator(paths.end());
-          result.insert(result.end(), begin, end);
-          std::sort(result.begin(), result.end());
-        }
-        break;
-      }
-    }
+  std::vector<std::filesystem::path> result;
+  std::error_code ec{};
+  auto dir = std::filesystem::recursive_directory_iterator(root_dir, ec);
+
+  if (ec)
+    return caf::make_error(ec::filesystem_error, ec.message());
+
+  auto begin = std::filesystem::begin(dir);
+  auto end = std::filesystem::end(dir);
+  result.reserve(std::distance(begin, end));
+  while (begin != end) {
+    const auto current_path = begin->path();
+    const auto current_depth = static_cast<size_t>(begin.depth());
+    if (current_depth >= max_recursion)
+      break;
```

Comment:
We should probably emit a warning when we hit the max recursion limit to let the user know that the result might not be accurate. (We didn't do this before, but sine we're refactoring this piece, now is the time.)

Reply:
We could also err with `ec::recursion_limit_reached`. @tenzir/backend ?

Reply:
Returning an error is always preferred since it allows us to print the error at the outermost place possible, and sometimes that allows us to print an error from the server in the client process.

Reply:
In the case of the recursion depth being reached, we now return an error with `ec::recursion_limit_reached` which should get printed from the callers when we call `error()` member function on the `caf::error` type returned from `filter_dir`.

How does that look?

Reply:
It's great!